File: 2024/1/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n")
  .map((a) => a.split("   ").map(Number));

const list1 = input.map((a) => a[0]);
const list2 = input.map((a) => a[1]);
list1.sort((a, b) => a - b);
list2.sort((a, b) => a - b);

const part1 = list1.reduce((acc, cur, i) => acc + Math.abs(cur - list2[i]), 0);
console.log(part1);

const count2 = new Map();
for (const n of list2) count2.set(n, (count2.get(n) || 0) + 1);

const part2 = list1.reduce((acc, cur) => acc + cur * (count2.get(cur) || 0), 0);
console.log(part2);

File: 2024/1/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strconv"
	"strings"

	"github.com/weitecklee/adventofcode/utils"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := strings.Split(string(data), "\n")
	list1, list2 := parseInput(&puzzleInput)
	fmt.Println(part1(list1, list2))
	fmt.Println(part2(list1, list2))
}

func parseInput(puzzleInput *[]string) (*[]int, *[]int) {
	list1 := make([]int, len(*puzzleInput))
	list2 := make([]int, len(*puzzleInput))
	numRegex := regexp.MustCompile(`^(\d+)\s+(\d+)$`)
	for i, s := range *puzzleInput {
		matches := numRegex.FindStringSubmatch(s)
		if n, err := strconv.Atoi(matches[1]); err != nil {
			panic(err)
		} else {
			list1[i] = n
		}
		if n, err := strconv.Atoi(matches[2]); err != nil {
			panic(err)
		} else {
			list2[i] = n
		}
	}
	sort.Ints(list1)
	sort.Ints(list2)
	return &list1, &list2
}

func part1(list1, list2 *[]int) int {
	res := 0
	for i := range *list1 {
		res += utils.AbsInt((*list1)[i] - (*list2)[i])
	}
	return res
}

func part2(list1, list2 *[]int) int {
	res := 0
	i := 0
	j := 0
	for i < len(*list1) && j < len(*list2) {
		for i < len(*list1) && (*list1)[i] < (*list2)[j] {
			i++
		}
		for j < len(*list2) && (*list1)[i] > (*list2)[j] {
			j++
		}
		for j < len(*list2) && (*list1)[i] == (*list2)[j] {
			res += (*list1)[i]
			j++
		}
	}
	return res
}
